File: String-Problems/Is-Palindrome/Is-Palindrome.cs
using System;
using System.Linq;

public class IsPalindromeSolution
{
    /// <summary>
    /// Returns true iff s is a palindrome.
    /// Clarification to ask the interviewer:
    /// - Are whitespace, punctuation, and case considered important?
    /// - For this function, we assume that they are NOT important.
    /// </summary>
    /// <param name="s">The candidate palindrome string.</param>
    /// <returns>True iff s is a palindrome.</returns>
    public static bool IsPalindrome(string s)
    {
        // Strip out all whitespace and punctuation, and ignore case.
        var strippedChars = s.ToCharArray()
                             .Where(c => !char.IsWhiteSpace(c)
                                         && !char.IsPunctuation(c))
                             .ToArray();
        var stripped = new string(strippedChars).ToLower();

        // Compare the first half of the string with the second half.
        var len = stripped.Length;
        for (var i = 0; i < len; i++)
        {
            if (stripped[i] != stripped[len - 1 - i])
            {
                return false;
            }
        }
        return true;
    }
    
    /// <summary>
    /// Returns the raw representation of s. Surrounds s with single quotes
    /// and allows us to see newlines and tabs.
    /// </summary>
    /// <param name="s">The string to be processed.</param>
    /// <returns>The raw representation of s.</returns>
    public static string ToRepr(string s)
    {
        return "'" + s.Replace("\n", @"\n").Replace("\t", @"\t") + "'";
    }

    public static void Main()
    {
        var testStrings = new[] {
			"",
			"#$%^%$#",
			"rAcecar",
			"racecar",
			"ricecar",
			"abba",
			"12344329",
			"a@@#  b C \n\t d Eedc b  A",
			"a@@#  b C \n\t d Eedc b  z",
			"A man, a plan, a canal, Panama!"
		};
        var testCase = 1;
        foreach (var s in testStrings)
        {
            Console.WriteLine("Test String #{0}:", testCase);
            Console.WriteLine("{0} {1} a palindrome.",
                              ToRepr(s), IsPalindrome(s) ? "is" : "is not");
            testCase++;
        }

        Console.Write("\nPress any key to continue...");
        Console.ReadKey();
    }
}

File: String-Problems/Is-Palindrome/Is-Palindrome.py
from string import punctuation


def is_palindrome(s):
    """
    Returns true iff the string s is a palindrome.
    Clarification to ask the interviewer:
    - Are whitespace, punctuation, and case considered important?
    - For this function, we assume that they are NOT important.
    """

    # Strip out all whitespace and punctuation, and ignore case.
    no_space_chars = list("".join(s.split()))
    no_punc_chars = [char for char in no_space_chars if char not in punctuation]
    stripped = "".join(no_punc_chars).lower()

    # Compare the first half of newS with the last half and return false
    # if any pair doesn't match. If len(newS) is odd, there is no need to look
    # at the middle character.
    # NOTE: Unlike '/', the '//' operator performs floor division.

    #---------------------------------------------------------------------------
    # Solution 1: Imperative.
    #---------------------------------------------------------------------------
    #    for i in range(len(stripped) // 2):
    #        if stripped[i] != stripped[len(stripped) - 1 - i]:
    #            return False
    #    return True
    #---------------------------------------------------------------------------
    # Solution 2: Declarative (negated quantifiers).
    #---------------------------------------------------------------------------
    #    return not any([stripped[i] != stripped[len(stripped) - 1 - i]
    #                    for i in range(len(stripped) // 2)])
    #---------------------------------------------------------------------------
    ## Solution 3: Declarative (normal quantifiers).
    #---------------------------------------------------------------------------
    return all([stripped[i] == stripped[len(stripped) - 1 - i]
                for i in range(len(stripped) // 2)])


if __name__ == "__main__":
    testStrings = [
        "",
        "#$%^%$#",
        "rAcecar",
        "racecar",
        "ricecar",
        "abba",
        "12344329",
        "a@@#  b C \n\t d Eedc b  A",
        "a@@#  b C \n\t d Eedc b  z",
        "A man, a plan, a canal, Panama!"
    ]
    for (index, s) in enumerate(testStrings):
        print(str.format("Test String #{0}:", index + 1))
        print(str.format("{0} {1} a palindrome.",
                         repr(s), "is" if is_palindrome(s) else "is not"))